main.py

```diff
@@ -47,35 +52,81 @@ def start_server(self, socket=os_socket):
                                 break
                             message = json.loads(data)
                             print()
-                            print(f"Received by {data}")
-                            logger.debug(f"Received by {message['sender']}: {message}")
-                            self.message_queue.put(data)
 
-                            if not self.message_queue.empty():
-                                msg = self.message_queue.get()
+                            if message.get("type") == "GET_NODES":
+                                conn.sendall(json.dumps({"nodes": list(self.peer_hosts)}).encode())
+                                self.peer_hosts.add(addr[0])
+                                print(f"server connected to {self.peer_hosts}")
+
+                                self.send_peers(peer_port)
+
+                            elif message.get("type") == "NEW_NODES":
+                                self.peer_hosts.update(message.get("nodes", []))
+                                self.peer_hosts.remove(self.ip)
+                            else:
+                                print(f"Received by {data}")
+                                logger.debug(f"Received by {message['sender']}: {message}")
+                                self.message_queue.put(data)
+
+                                if not self.message_queue.empty():
+                                    msg = self.message_queue.get()
 
-                                ack = json.dumps({"type": "ack", "message": f"Received {message['message']} from {message['sender']}", "sender": self.nickname}).encode()
-                                conn.sendall(ack)
-                                logger.debug(f"{self.nickname} sent ack {ack}")
+                                    ack = json.dumps({"type": "ack", "message": f"Received {message['message']} from {message['sender']}", "sender": self.nickname}).encode()
+                                    conn.sendall(ack)
+                                    logger.debug(f"{self.nickname} sent ack {ack}")
         except Exception as exc:
             logger.exception(exc)
             raise exc
 
+    def request_peers(self, peer_port, socket=os_socket):
+        try:
+            with socket(AF_INET, SOCK_STREAM) as s:
+                s.connect((list(self.peer_hosts)[0], peer_port))
+                s.sendall(json.dumps({"type": "GET_NODES"}).encode())
+                data = s.recv(1024)
+
+                response = json.loads(data)
+
+        except Exception as exc:
+            logger.error(f"Failed to connect to startup server: {exc}")
```

Comment:
You need to `raise exc` after this as the rest of the code cannot work if response was not received